From f6b2c3571ea70d1c86be049ead0aeb80f32c0ab6 Mon Sep 17 00:00:00 2001
From: Joni Poikelin <joni.poikelin@qt.io>
Date: Mon, 15 Mar 2021 12:54:45 +0200
Subject: [PATCH 173/182] Fix removing columns when QSortFilterProxyModel has
 not been sorted yet

Fixes: QTBUG-91788
Change-Id: Iddcafd3171f0f3703b94893a32b4ccaaeea9e713
Reviewed-by: David Faure <david.faure@kdab.com>
(cherry picked from commit e287582cba74ddcfc67f83647e72c0b23bd99c36)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
(cherry picked from commit 6bd89d122e488d83c3e9310cf102ef238de8ac00)
---
 src/corelib/itemmodels/qsortfilterproxymodel.cpp   |  5 ++++-
 .../tst_qconcatenatetablesproxymodel.cpp           | 14 ++++++++++++++
 2 files changed, 18 insertions(+), 1 deletion(-)

diff --git a/src/corelib/itemmodels/qsortfilterproxymodel.cpp b/src/corelib/itemmodels/qsortfilterproxymodel.cpp
index 9fc59d5894..1abdafcdbe 100644
--- a/src/corelib/itemmodels/qsortfilterproxymodel.cpp
+++ b/src/corelib/itemmodels/qsortfilterproxymodel.cpp
@@ -1802,7 +1802,10 @@ void QSortFilterProxyModelPrivate::_q_sourceColumnsRemoved(
             source_sort_column = -1;
     }
 
-    proxy_sort_column = q->mapFromSource(model->index(0,source_sort_column, source_parent)).column();
+    if (source_sort_column >= 0)
+        proxy_sort_column = q->mapFromSource(model->index(0,source_sort_column, source_parent)).column();
+    else
+        proxy_sort_column = -1;
 }
 
 void QSortFilterProxyModelPrivate::_q_sourceColumnsAboutToBeMoved(
diff --git a/tests/auto/corelib/itemmodels/qconcatenatetablesproxymodel/tst_qconcatenatetablesproxymodel.cpp b/tests/auto/corelib/itemmodels/qconcatenatetablesproxymodel/tst_qconcatenatetablesproxymodel.cpp
index 40617c1f7d..b3e46beafb 100644
--- a/tests/auto/corelib/itemmodels/qconcatenatetablesproxymodel/tst_qconcatenatetablesproxymodel.cpp
+++ b/tests/auto/corelib/itemmodels/qconcatenatetablesproxymodel/tst_qconcatenatetablesproxymodel.cpp
@@ -116,6 +116,7 @@ private Q_SLOTS:
     void shouldPropagateDropBetweenItemsAtModelBoundary();
     void shouldPropagateDropAfterLastRow_data();
     void shouldPropagateDropAfterLastRow();
+    void qtbug91788();
 
 private:
     QStandardItemModel mod;
@@ -818,6 +819,19 @@ void tst_QConcatenateTablesProxyModel::shouldPropagateDropAfterLastRow()
 
 }
 
+void tst_QConcatenateTablesProxyModel::qtbug91788()
+{
+    QConcatenateTablesProxyModel proxyConcat;
+    QStringList strList{QString("one"),QString("two")};
+    QStringListModel strListModelA(strList);
+    QSortFilterProxyModel proxyFilter;
+    proxyFilter.setSourceModel(&proxyConcat);
+
+    proxyConcat.addSourceModel(&strListModelA);
+    proxyConcat.removeSourceModel(&strListModelA); // This should not assert
+    QCOMPARE(proxyConcat.columnCount(), 0);
+}
+
 QTEST_GUILESS_MAIN(tst_QConcatenateTablesProxyModel)
 
 #include "tst_qconcatenatetablesproxymodel.moc"
-- 
2.31.1

